### ISSUE TITLE
RunOnceAtStart sometimes it doesn't work.

### ISSUE THREAD

**Issue by Kerwin1202** [jamesmh/coravel#353]
**Describe the bug**
```
    scheduler.Schedule<Service>()
        .Cron("30 */2 * * *")
        .RunOnceAtStart()
        .Zoned(TimeZoneInfo.Local);

    scheduler.ScheduleAsync(async () =>
        {
             Console.WriteLine("Start refresh data");
           // ... other code 
        })
        .DailyAtHour(9)
        .RunOnceAtStart()
        .Zoned(TimeZoneInfo.Local);
```
```
public class Service : IInvocable
{
   ......  other code
    public async Task Invoke()
    {
        _logger. Information($"Start login {_config. UserName}");
        await Login();
    }
   ......  other code
}
```

**Affected Coravel Feature**
Scheduling

**Expected behaviour**
Output content immediately when starting the program. ( "Start login xxx" and "Start refresh data") , but when I run it many times, sometimes it does not output content, and the console does not report any error message.

run windows service. 
.net core 7.0
coravel : 5.0.2.0

**Comment by sneakpodbob** [comment]
I have the exact same issue on .NET8.

`app.Services.UseScheduler(scheduler =>
{
    scheduler.Schedule<CostInvocable>()
            .EveryThirtySeconds()
            .PreventOverlapping(nameof(CostInvocable));

    scheduler.Schedule<PbsInvocable>()
        .DailyAtHour(3)
        .RunOnceAtStart()
        .PreventOverlapping(nameof(PbsInvocable));
});`

Both Invocables have a log-message as the first thing they do - the upper one starts a couple of seconds after launch, the second one never shows up..

**Comment by arteny** [comment]
Not works for me in .net 8 with ScheduleAsync as well

**Comment by chris-rp** [comment]
This is still an issue in
.net8, coravel 5.0.4

Anything new on this issue?

**Comment by jamesmh** [comment]
No news on this yet. Will look to investigate soon. Not sure what's going on at first glance. Unless someone else has any idea by looking at the source?

**Comment by jamesmh** [comment]
Curious for those who reported this: what kind of machine was this on? A small container? Local machine with many threads?

I have an idea of a potential race condition causing this, but can't reproduce at all ATM.

**Comment by jamesmh** [comment]
Was able to reproduce via unit testing the suspect concurrency scenario. Fixed in 6.0.2!

**Comment by jamesmh** [comment]
Please re-open if you find the issue persists. It's still possible that the issue has some other causes related to the .NET lifecycle 🤷.

**Comment by arteny** [comment]
Still not works on my side with 6.0.2
from this breakpoint:
```
		Scheduler.ScheduleAsync(async () => { await updateExposure(); })
			.EveryFifteenSeconds()
			.RunOnceAtStart();
```
desitnation method is called after several seconds

**Comment by jamesmh** [comment]
🥲 Thanks @arteny - I'll have to take another gander at this.

**Comment by jamesmh** [comment]
@arteny It's possible that there's another task causing this one to be delayed. If you put this task first in the list of all schedule tasks, does it improve things?

**Comment by jamesmh** [comment]
For example:

```csharp
                scheduler.Schedule(() => Thread.Sleep(10000)).EverySecond();
                scheduler.ScheduleAsync(async () => { await Task.Delay(1); Console.WriteLine("SHOULD GO FIRST"); })
                    .Hourly()
                    .RunOnceAtStart();
```

Because the first task is not returning a `Task` (e.g. not `await/async`) it hogs the thread. So the second task won't run until the first one completes (after 10 seconds....). So it "looks" like the second one doesn't run immediately upon start up, but it IS triggered properly. It's just that another task is not letting go of the thread.

That's what schedule workers are for. So this works:

```csharp
                scheduler.Schedule(() => Thread.Sleep(10000)).EverySecond();
                scheduler.OnWorker("TEST");
                scheduler.ScheduleAsync(async () => { await Task.Delay(1); Console.WriteLine("SHOULD GO FIRST"); })
                    .Hourly()
                    .RunOnceAtStart();
```

This code will put the second scheduled task on it's own thread, so it's not blocked / waiting on the first task to complete.

It might be the case that everyone or some of you guys/gals are experiencing this issue. Otherwise, I cannot reproduce this issue at all, other than this approach.

So what I recommend for everyone having this issue is to either list these "run once at start" tasks first in your configuration (so they run first at run time) or put them on a dedicated worker.

Let me know this works for anyone or doesn't! Thanks!

**Comment by arteny** [comment]
@jamesmh No, I have only one task for scheduler which is in above example.
And issues still reproduced. RunOnceAtStart not works.
Maybe it is related some way that I am running it in Blazor app.
Or I incorrectly understood RunOnceAtStart behavior. I am adding my scheduler after app.Run() call in program workflow.
And interval scheduling working well, I think it will be usefult to add some method to call schedule action right after adding it with ScheduleAsync whenever it is added.
Now I am using such extension method for this:
```cs
	public static IScheduleInterval RunAndSchedule(this IScheduler scheduler, Func<Task> task)
	{
		task.Invoke();
		return scheduler.ScheduleAsync(task);
	}
```
But anyway, looks like library not for this type scenario, because it has no stop scheduler methods.

**Comment by Kerwin1202** [comment]
I’m very sorry. Today, after reviewing my code again, I found that Coravel actually works fine.

Sometimes it didn’t execute because I had two RunOnceAtStart. I didn’t show all of them before. In the second RunOnceAtStart, which always executes, there was code that blocked the main thread, similar to the following code.


```c#
scheduler.ScheduleAsync(async () =>
{
    // this code  sometimes output
    Console.WriteLine($"{DateTime.Now} this code  sometimes output ...");
    try
    {
        // my code
    }
    catch (Exception e)
    {
        Console.WriteLine(e.ToString());
    }
})
.DailyAtHour(9)
.RunOnceAtStart()
.PreventOverlapping("11")
.Zoned(TimeZoneInfo.Local);

scheduler.ScheduleAsync(async () =>
{
    // if deployed server, will goin here
    // my code 
    // this code  always output in server 
    Console.WriteLine($"{DateTime.Now} this code  always output ...");
    await Task.Delay(10000);
    //  My actual code is like The following code will cause the scheduler to be stuck here and not execute any subsequent scheduled tasks.
    // Because this code is a dead loop, it will never exit the execution.
    // await Task.Run(() =>
    // {
    //     while (true)
    //     {
    //         Task.Delay(1000).Wait();
    //     }
    // });
})
.Weekly().RunOnceAtStart()
.PreventOverlapping("7")
.Zoned(TimeZoneInfo.Local);
```

So I tried commenting out the while loop and replaced it with await Task.Delay(10000);.
I found that the first RunOnceAtStart would execute with a delay, producing output similar to the following.
```
First run :
2025/9/1 23:38:40 this code  always output ...
2025/9/1 23:38:50 this code  sometimes output ...
^Cinfo: Microsoft.Hosting.Lifetime[0]
      Application is shutting down...


Sencond run:
2025/9/1 23:39:02 this code  sometimes output ...
2025/9/1 23:39:02 this code  always output ...
^Cinfo: Microsoft.Hosting.Lifetime[0]
```


Because my two RunOnceAtStart are triggered at different times, I can’t put the code from both into the same RunOnceAtStart.

```
.Weekly().RunOnceAtStart()

.DailyAtHour(9).RunOnceAtStart()
```

After I commented out await Task.Delay(10000);, I ran it more than 20 times, and each time both RunOnceAtStart executed and produced the expected output. Therefore, I believe Coravel works perfectly well.


BTW:  .net8, coravel 6.0.2


But I think the normal expectation for ScheduleAsync is that, regardless of RunOnceAtStart, it should be triggered in parallel using multiple threads. If this could be adjusted, that would be ideal.

**Comment by jamesmh** [comment]
@Kerwin1202 thanks for the update! I suspect that's what's happening in most instances - something is causing the one task/job you expect to run immediately to wait until the other one(s) are completed.

Agree that there's an approach to using multiple threads. For anyone on this thread you CAN do that by using [schedule workers](https://docs.coravel.net/Scheduler/#schedule-workers):

```csharp
scheduler.OnWorker("CPUIntensiveTasks");
scheduler
    .Schedule<RebuildStaticCachedData>().Hourly().RunOnceAtStart();
```

If you wrap the job you want to start immediately on start, see if that helps. If it does, then it means another jobs is hogging the schedule thread before allowing any other jobs to run.
